x-pack/filebeat/input/shipper/input.go

```diff
@@ -0,0 +1,279 @@
+		err := stream.client.Close()
+		if err != nil {
+			in.log.Infof("error closing client for stream: %s: %w", streamID, stream)
+		}
+	}
+	in.srvMut.Lock()
+	defer in.srvMut.Unlock()
+	if in.shipper != nil {
+		err := in.shipper.Close()
+		if err != nil {
+			in.log.Debugf("Error stopping shipper input: %s", err)
+		}
+		in.shipper = nil
+	}
+	if in.server != nil {
+		in.server.GracefulStop()
+		in.server = nil
+
+	}
+	err := os.Remove(in.shipperSrv)
+	if err != nil {
+		in.log.Debugf("error removing unix socket for grpc listener during shutdown: %s", err)
+	}
+}
+
+// Run the shipper
+func (in *shipperInput) Run(inputContext v2.Context, pipeline beat.Pipeline) error {
+	in.log.Infof("Running shipper input")
+	// create clients ahead of time
+	for streamID, streamProc := range in.streams {
+		client, err := pipeline.ConnectWith(beat.ClientConfig{
+			// TODO: need an EventListener?
+			Processing: beat.ProcessingConfig{
+				Processor: streamProc.processors,
+			},
+			CloseRef: inputContext.Cancelation,
+		})
+		if err != nil {
+			return fmt.Errorf("error creating client for stream %s: %w", streamID, err)
+		}
+		in.log.Infof("Creating beat client for stream %s", streamID)
+
+		newStreamData := streamData{client: client, index: in.streams[streamID].index, processors: in.streams[streamID].processors}
+		in.streams[streamID] = newStreamData
+	}
+
+	//setup gRPC
+	err := in.setupgRPC(pipeline)
+	if err != nil {
+		return fmt.Errorf("error starting shipper gRPC server: %w", err)
+	}
+	in.log.Infof("done setting up gRPC server")
+
+	// wait for shutdown
+	<-inputContext.Cancelation.Done()
+
+	in.Stop()
+
+	return nil
+}
+
+func (in *shipperInput) setupgRPC(pipeline beat.Pipeline) error {
+	in.shipperSrv = strings.TrimPrefix(in.cfg.Conn.Server, "unix://")
+	in.log.Infof("initializing grpc server at %s", in.shipperSrv)
+	// Currently no TLS until we figure out mTLS issues in agent/shipper
+	creds := insecure.NewCredentials()
+	opts := []grpc.ServerOption{
+		grpc.Creds(creds),
+		grpc.MaxRecvMsgSize(64 * units.MiB),
+	}
+
+	var err error
+	in.server = grpc.NewServer(opts...)
+	in.shipper, err = NewShipperServer(pipeline, in)
+	if err != nil {
+		return fmt.Errorf("error creating shipper server: %w", err)
+	}
+
+	pb.RegisterProducerServer(in.server, in.shipper)
+
+	in.srvMut.Lock()
+
+	// paranoid checking, make sure we have the base directory.
+	dir := filepath.Dir(in.shipperSrv)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0o755)
+		if err != nil {
+			in.srvMut.Unlock()
+			return fmt.Errorf("could not create directory for unix socket %s: %w", dir, err)
+		}
+	}
+
+	// on linux, net.Listen will fail if the file already exists
+	if _, err = os.Stat(in.shipperSrv); err == nil {
+		in.log.Debugf("listen address %s already exists, removing", in.shipperSrv)
+		err = os.Remove(in.shipperSrv)
+		if err != nil {
+			in.srvMut.Unlock()
+			return fmt.Errorf("error removing unix socket %s: %w", in.shipperSrv, err)
+		}
+	}
```

Comment:
This has a few race conditions in it.  For example the directory could be removed after you make it, but before the shipper starts.  Or file could be replaced after you delete it but before the server starts.  So a better plan is to do the `os.MkdirAll` and log any errors but proceed.  Also try the `os.Remove` and log any error that isn't `errNotExist`.  But still try the `newListener`.  The idea is, we try to provide a clean environment, but we trust `newListener` to error if somehow the environment isn't clean when we start.

Reply:
Ah, good point, yah.